chromium/browser_type.go

```diff
@@ -60,9 +60,70 @@ func NewBrowserType(vu k6modules.VU) api.BrowserType {
 
 // Connect attaches k6 browser to an existing browser instance.
 func (b *BrowserType) Connect(wsEndpoint string, opts goja.Value) api.Browser {
-	rt := b.vu.Runtime()
-	k6common.Throw(rt, errors.New("BrowserType.connect() has not been implemented yet"))
-	return nil
+	ctx := b.initContext()
+
+	var err error
+	var logger *log.Logger
+	if logger, err = makeLogger(ctx); err != nil {
+		k6ext.Panic(ctx, "setting up logger: %w", err)
+	}
+	launchOpts := common.NewLaunchOptions(k6ext.OnCloud(), true)
+	if err := launchOpts.Parse(ctx, logger, opts); err != nil {
+		k6ext.Panic(ctx, "parsing launch options: %w", err)
+	}
+	ctx = common.WithLaunchOptions(ctx, launchOpts)
+
+	bp, err := b.connect(ctx, wsEndpoint, launchOpts, logger)
+	if err != nil {
+		err = &k6ext.UserFriendlyError{
+			Err:     err,
+			Timeout: launchOpts.Timeout,
+		}
+		k6ext.Panic(ctx, "%w", err)
+	}
+
+	return bp
+}
+
+func (b *BrowserType) connect(ctx context.Context, wsURL string, opts *common.LaunchOptions, logger *log.Logger) (*common.Browser, error) {
+	if err := logger.SetCategoryFilter(opts.LogCategoryFilter); err != nil {
+		return nil, fmt.Errorf("%w", err)
+	}
+	if opts.Debug {
+		_ = logger.SetLevel("debug")
+	}
+
+	browserProc, err := b.link(ctx, wsURL, opts, logger)
+	if browserProc == nil {
+		return nil, fmt.Errorf("connecting to browser: %w", err)
+	}
+
+	// If this context is cancelled we'll initiate an extension wide
+	// cancellation and shutdown.
+	browserCtx, browserCtxCancel := context.WithCancel(ctx)
+	b.Ctx = browserCtx
+	browser, err := common.NewBrowser(
+		browserCtx, browserCtxCancel, browserProc, opts, logger,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("connecting to browser: %w", err)
+	}
+
+	return browser, nil
+}
+
+func (b *BrowserType) link(
+	ctx context.Context, wsURL string,
+	opts *common.LaunchOptions, logger *log.Logger,
+) (_ *common.BrowserProcess, rerr error) {
+	bProcCtx, bProcCtxCancel := context.WithTimeout(ctx, opts.Timeout)
+	defer func() {
```

Comment:
I'm not sure i understand why we need to defer this, could we check the error straight after `NewRemoteBrowserProcess`?

Reply:
I kept this to mimic the way `launch` [implements](https://github.com/grafana/xk6-browser/blob/e24732bd0a56656bc7ea885d305847c83a640f81/chromium/browser_type.go#L242-L247) it. I guess this was done like that in order to not be able to handle explicitly the error returned from `NewBrowserProcess`, and just return directly with the call to the method. So, no strong opinion here, I thing the defer makes it a little bit more complicated IMO. Should we change both implementations?

Reply:
Yes, i agree, the defer makes it slightly more complicated. I would prefer a simpler approach.

Reply:
The `defer` ensures that we cancel the context, and I'm on the side of not removing it and leaving it as it is.

Reply:
Do you prefer this:
```go
bProcCtx, bProcCtxCancel := context.WithTimeout(ctx, opts.Timeout)
p, err := common.NewRemoteBrowserProcess(bProcCtx, wsURL, bProcCtxCancel, logger)
if err != nil {
    bProcCtxCancel()
}

return p, err //nolint: wrapcheck
```
Or this:
```go
bProcCtx, bProcCtxCancel := context.WithTimeout(ctx, opts.Timeout)
p, err := common.NewRemoteBrowserProcess(bProcCtx, wsURL, bProcCtxCancel, logger)
if err != nil {
    bProcCtxCancel()
    return nil, err //nolint:wrapcheck
}

return p, nil
```
I think this last one is probably more idiomatic, but no strong opinions and first one is 1 line less :)

Reply:
The idiomatic one is the `defer` approach :)

Reply:
@inancgumus I think the defer in the `launch` [implementation](https://github.com/grafana/xk6-browser/blob/e24732bd0a56656bc7ea885d305847c83a640f81/chromium/browser_type.go#L242-L247) could be left as is since there are multiple failure paths, but the defer here makes it a little more complicated. There's only one place where the error could be non-nil, which is after `NewRemoteBrowserProcess` is called, so it would make things simpler if we just checked the error there.

Happy to leave it as though if there's a strong opinion on keeping defer.

Reply:
> The idiomatic one is the `defer` approach :)

Really? I find it quite complicated to grasp on first sight :sweat_smile:

Reply:
@ankur22, ah, right, `defer` is unnecessary there 😅

Reply:
Done, see [here](https://github.com/grafana/xk6-browser/blob/e1ca87ca0963b2528aaf4d8f1c01e0fcfc39031a/chromium/browser_type.go#L141-L153).